### PR TITLE
Websocket implementation of a whiteboard

### LINKED ISSUE #2: Create a more interesting example for the HTTP/2 Cloud Run Fastify article
The article [being written](https://docs.google.com/document/d/1yMtFNnS6jpNnb4Wv8u2r5ussDeAmDYG649-QchPG-9g/edit#)

The card where the article activity is tracked is [this](https://github.com/orgs/nearform/projects/3#card-53783968).

### PR DESCRIPTION
**Pull request by Paul-Isache**
1. Create simple whiteboard that draws lines
2. Connect to websocket 
3. Draw received events

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

File: ws/public/main.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        state.color = event.target.getAttribute('data-color');
+        event.target.classList.add('active')
+    }, false)
+
+    window.addEventListener('resize', setCanvas, false)
+    setCanvas()
+
+    document.getElementsByClassName('color')[0].classList.add('active')
+    state.color = document.getElementsByClassName('active')[0].getAttribute('data-color');
+})()
```
Comment (simoneb) [inline ws/public/main.js]: nit: new line at the end of the file

File: ws/index.js [hunk 2]
```diff
@@ -9,4 +9,16 @@ fastify.register(require('fastify-websocket'), {
...
 
+fastify.get('/', {
+  websocket: true
+}, (connection /* SocketStream */ , req /* FastifyRequest */ ) => {
+  connection.socket.on('message', message => {
+    fastify.websocketServer.clients.forEach((client) => {
+      if (client.readyState === 1) {
+        client.send(message)
+      }
+    })
```
Comment (simoneb) [inline ws/index.js]: question: are we sending this to the sender too?

File: ws/public/main.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+(function () {
+    const canvas = document.getElementsByClassName('whiteboard')[0]
```
Comment (simoneb) [inline ws/public/main.js]: please use 2 spaces for indentation across this file and everywhere else

Comment (simoneb) [inline ws/public/main.js]: nit: I would find it more conceptually correct that we reference this element by id, as there is only one ever supposed to exist in the app

File: ws/public/style.css [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    background-color: green;
+}
+
+.color[data-color='blue'] {
+    background-color: blue;
+}
+
+.color[data-color='yellow'] {
+    background-color: yellow;
+}
```
Comment (simoneb) [inline ws/public/style.css]: new line at end of file

File: ws/public/main.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    canvas.addEventListener('mousemove', onMouseMove, false)
+
+    //Touch support for mobile devices
+    canvas.addEventListener('touchstart', onMouseDown, false)
+    canvas.addEventListener('touchend', onMouseUp, false)
+    canvas.addEventListener('touchcancel', onMouseUp, false)
+    canvas.addEventListener('touchmove', onMouseMove, false)
+
+
+    document.getElementsByClassName('colors')[0].addEventListener('click', (event) => {
```
Comment (simoneb) [inline ws/public/main.js]: here too, use an id when you're referencing an element in the DOM of which you know there will only ever exist 1 occurrence

File: ws/public/main.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        if (event.target.className.includes('colors')) {
+            return
+        }
+
+        document.getElementsByClassName('active')[0].classList.remove('active')
+
+        state.color = event.target.getAttribute('data-color');
+        event.target.classList.add('active')
```
Comment (simoneb) [inline ws/public/main.js]: I find this logic could be written in a simpler way. e.g. you're adding an event listener on the parent but then you're returning early if it's the parent that was clicked. also getting a list of elements to always get the first one doesn't feel right and suggest that's a better way to write this

File: ws/public/main.js (hunk 1 above)
Comment (simoneb) [reply ws/public/main.js]: I would find it more idiomatic if you weren't executing this function as a iife but if it was a DOMContentLoaded event listener

File: ws/public/main.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        context.beginPath()
+        context.moveTo(from.x, from.y)
+        context.lineTo(to.x, to.y)
+        context.strokeStyle = color
+        context.lineWidth = 2
+        context.stroke()
+        context.closePath()
+    }
+
+    const draw = (emit = false) => {
```
Comment (simoneb) [inline ws/public/main.js]: unless I'm missing something, this function is only ever called with emit = true, so I'm not sure what this parameter is used for

**Review by simoneb — CHANGES_REQUESTED** [review]
See inline comments. Also, let's have a standalone example for this, not in the same folder as the existing example.

The idea is that we show a set of simple examples, in the blog post, then we conclude with extra content which is the whiteboard, which is standalone and should mimic what's already being done in the other examples, with its own readme and so on

Also, don't add .DS_Store, add it to .gitignore

File: ws/index.js (hunk 2 above)
Comment (Paul-Isache) [reply ws/index.js]: for the purpose of this i was sending it to the sender as well. but i can try to filter the clients in some manner by session maybe by adding session to it.

Do you want me to add this?

File: ws/public/main.js (hunk 7 above)
Comment (Paul-Isache) [reply ws/public/main.js]: it was used but then i refactored the code and extracted the first part in a function. I will remove it

File: ws/public/main.js (hunk 5 above)
Comment (Paul-Isache) [reply ws/public/main.js]: I replaced it with an id instead of class

File: ws/public/main.js (hunk 3 above)
Comment (Paul-Isache) [reply ws/public/main.js]: I used id now instead. for me the id seems very "powerful" to use

Comment (Paul-Isache) [reply ws/public/main.js]: ok

File: ws/public/main.js (hunk 6 above)
Comment (Paul-Isache) [reply ws/public/main.js]: Sorry but i don't understand the comment.
I am adding the event on the parent and checking if the click was on one of the colors. The list of active should consist one single element being active at any time.

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1,3 @@
 node_modules/
-package-lock.json
\ No newline at end of file
+package-lock.json
+.DS_STORE
```
Comment (simoneb) [inline .gitignore]: please configure your editor to always add a new line at the end of each file

File: examples/whiteboard/index.html [hunk 9]
```diff
@@ -0,0 +1,24 @@
+<!doctype html>
+<html lang="en">
+
+<head>
+  <meta charset="UTF-8">
+  <title>Socket implementation</title>
```
Comment (simoneb) [inline examples/whiteboard/index.html]:
```suggestion
  <title>Fastify WebSocket Whiteboard</title>
```

File: ws/index.js (hunk 2 above)
Comment (simoneb) [reply ws/index.js]: aren't we drawing twice if we're not excluding the sender from this loop?

File: .gitignore (hunk 8 above)
Comment (simoneb) [reply .gitignore]: you still don't have a new line at the end of the file

File: examples/whiteboard/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Fastify WebSocket Whiteboard
```
Comment (simoneb) [inline examples/whiteboard/README.md]: please inspire from the other readmes and show instructions about how to run this in cloud run too. you must be able to run the whole app in cloud run, meaning, serve the static assets from there, for istance using fastify-static.

Comment (Paul-Isache) [reply examples/whiteboard/README.md]: okey.. i wasn't aware of that. i will look how to serve static content using fastify

Comment (simoneb) [reply examples/whiteboard/README.md]: this repo is accompanying the article about fastify on google cloud run, so all examples should be able to be deployed and run there

File: whiteboard/index.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  connection.socket.on('message', message => {
+    fastify.websocketServer.clients.forEach((client) => {
+      if (client.readyState === 1 && client !== connection.socket) {
+        client.send(message)
+      }
+    })
+  })
+})
+
+fastify.listen(process.env.PORT || 8080)
```
Comment (simoneb) [inline whiteboard/index.js]:
```suggestion
fastify.listen(process.env.PORT || 3000)
```

File: whiteboard/index.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+const fastify = require('fastify')({
+    logger: true,
+})
+const path = require('path')
+
+fastify.register(require('fastify-static'), {
+    root: path.join(__dirname, 'public'),
+    prefix: '/'
```
Comment (simoneb) [inline whiteboard/index.js]: remove this

File: whiteboard/public/main.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
+'use strict';
+
+const WS_URL = 'ws://0.0.0.0:3000/ws'
```
Comment (simoneb) [inline whiteboard/public/main.js]: 1. can't be the hardcoded hostname and port, get it from location.origin (or something similar)
2. the scheme must reflect the http(s) scheme:
  - if http -> ws
  - if https -> wss

File: whiteboard/index.js (hunk 11 above)
Comment (simoneb) [reply whiteboard/index.js]: also note that you should specify the address as the second argument to .listen. see how the other examples do it

File: whiteboard/public/main.js (hunk 13 above)
Comment (Paul-Isache) [reply whiteboard/public/main.js]: Done

File: whiteboard/README.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    **Note:** If you have deployed on Cloud Run, you can replace `localhost:3000` with your Cloud Run service URL.
+
+1. After the server starts, make a request using `websockat` or another Web socket client.
+
+   ```sh
+   websocat ws://localhost:3000/ws
+   ```
+    
+    **Note:** If you have deployed on Cloud Run, you can replace `localhost:3000` with your Cloud Run service URL and use an online Web socket client like
```
Comment (simoneb) [inline whiteboard/README.md]: these instructions are wrong. the whiteboard example is not tested with these instructions but by using the UI instead

File: whiteboard/index.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  connection.socket.on('message', message => {
+    fastify.websocketServer.clients.forEach((client) => {
+      if (client.readyState === 1 && client !== connection.socket) {
+        client.send(message)
+      }
+    })
+  })
+})
+
+fastify.listen(process.env.PORT || 3000)
```
Comment (simoneb) [inline whiteboard/index.js]: this won't work. see how the other examples do it

**Review by simoneb — APPROVED** [review]

**Pull request merged**